csharp/src/SeedLang/Visualization/Events.cs

```diff
@@ -128,6 +131,31 @@ public class SingleStep : AbstractEvent {
       public SingleStep(TextRange range) : base(range) { }
     }
 
+    // An event which is triggered when a value is assigned to an element of a container.
+    //
+    // The container might be unnamed variables in following cases. The name of the container is
+    // null and the type is undefined in these cases.
+    // 1) Sets the element of a temporary container: [1, 2, 3][1] = 5
+    // 2) Sets the element of a intermediate container: a[1][2] = 5
```

Comment:
This is very tricky for the 2-dimensional matrix access, such as:

```
matrix2x3 = [ [1, 2], [3, 4], [5, 6]]
matrix2x3[1][1] = 9
```

This is a common case in basic algorithms or data structure programs.

The current implementation only outputs the following event:

```
SubscriptAssign: [3, 9][1] = 9
```

What's the meaning of `3, 9` here?

It's hard for the visualizer to tell which variable is accessed.

Is it possible to divide the operation `matrix2x3[1][1] = 9` into two phases:

- phase 1: get the left value by accessing the memory position of `matrix2x3[1]`, and assign the position to an `unnamed` variable.
- phase 2: assign `9` to the `unnamed` variable.

And send out the info of both the two phases to the client side?

As for a way more complex case, like:

```
matrix[3][4][a][b][c][9][8] = 10
```

It's kind of a chained memory access operation. Can we send out the full memory access chain out so that the client understands where the value is actually put in?

This is really a complex problem. We might need more considerations and experiments here.

Also consider how a typical modern debugger/insepctor handles the code line:

```
matrix[3][4][a][b][c][9][8] = 10
```

Does a modern debugger/inspector know where in the matrix list an internal element is being changed?

Reply:
New changes for this PR:
1. Refactor VariableResolver class to support chained names of variables. Fixed #151 
2. Remove block scope temporarily to make variable scoping is similar with Python.
3. Try to collect all information of multi-dimensional matrix, but found these information cannot be collected during compile time. It can only be collected during runtime. So only 1-dimensional array elements assignment notification is supported in this PR (Temporary array is also ignored). I will add an issue to consider how to track the reference of all array elements based on registers information during runtime.
4. From this conversation, I think maybe the most interested things of SeedLang clients are changes related named variables and elements of containers. So we can only focus on these changes, e.g. comparison of named variables and container elements, send notification of them, and ignore comparison of const of temporary variables. Is it reasonable?
5. Add `FluentAssertions` into unit test project to make testing easier.

Please have a review again.

Reply:
> Try to collect all information of multi-dimensional matrix, but found these information cannot be collected during compile time. It can only be collected during runtime. So only 1-dimensional array elements assignment notification is supported in this PR (Temporary array is also ignored). I will add an issue to consider how to track the reference of all array elements based on registers information during runtime.

This makes sense. Thanks!

> From this conversation, I think maybe the most interested things of SeedLang clients are changes related named variables and elements of containers. So we can only focus on these changes, e.g. comparison of named variables and container elements, send notification of them, and ignore comparison of const of temporary variables. Is it reasonable?

I agree. No need to care about temporary variables.

The most significant request from SeedLang users would be monitoring the change of a variable or an element of a list/dict variable.

But, there are still complex things even if we simplify things like this. Here is one example:

```
a = [1, 2, 3]
b = a
b[2] = 10
```

How can the visualizer know the content of the list `a` has been changed when it receives the assignment notification about `b[2] = 10`?

Let's finish this PR. We can discuss those more complex problems later. I will also create a new issue to track the a/b problem as showed above.